Applications/FileManager/FileUtils.cpp

```diff
@@ -9,6 +9,42 @@
 
 namespace FileUtils {
 
+int delete_directory(String directory, String& error_file)
+{
+    CDirIterator iterator(directory, CDirIterator::SkipDots);
+    if (iterator.has_error()) {
+        error_file = directory;
+        return -1;
+    }
+
+    while (iterator.has_next()) {
+        auto file_to_delete = String::format("%s/%s", directory.characters(), iterator.next_path().characters());
+        struct stat st;
+
+        if (lstat(file_to_delete.characters(), &st)) {
+            error_file = file_to_delete;
+            return errno;
```

Comment:
I don't think this should be returning `errno`; a better convention is to return `-1` and *set* `errno`. And in this case `lstat()` already sets `errno` for us.

Reply:
I agree, that's how I first tried it. But after calling `delete_directory` in `main.cpp` it seems that `errno` was overwritten and `strerror` always returned "Success (not an error)". I also tried to save `errno` before showing the message box, but no luck. I just checked again and the problem is the following:
When reading `errno` after the recursive call to `delete_directory` failed, `errno` is zero, even though it should have been set. I checked and confirmed that `errno` was set, but somehow returning from the failed `delete_directory` call reset `errno`. I honestly can't understand how this can happen. Here is the modified code I tested with:
```
            if (delete_directory(file_to_delete, file_to_delete)) { //This should fail, because file_to_delete will be a file we don't have permissions to delete
                int s = errno; //errno is 0, even though it should have been setto 13
                dbg() << String::format("%d\n", s); //This prints 0
                error_file = file_to_delete;
                return -1;
            }
        } else if (unlink(file_to_delete.characters())) { //This is where the recursive call from above will fail
            int s = errno;
            perror("unlink"); //This prints the proper error string "no permisson", and errno is properly set to 13
            error_file = file_to_delete;
            errno = s; //perror will reset errno, so I made sure to save it and set it again
            return -1;
        }
```

Reply:
`errno` will be overwritten by the `CDirIterator` destructor.
Trying to communicate failure through `errno` in the presence of destructors is very error-prone. :)

Reply:
Interesting, I did not even think about the destructor.  What would be the best way of dealing with that? Save the `errno` in the destructor before calling `closedir` and then setting it again?

Reply:
Nah, let's not try to set `errno` here. `errno` is LibC's thing, we don't need to mess with it.
I think it's okay to return `errno` like you're doing.

There's the `AK::Result` and `AK::Error` classes that @rburchell introduced a long time ago but we haven't gotten around to actually using. They might be interesting for this sort of thing. But there's no established idioms around their use yet.

Meanwhile I think this patch is actually okay as-is :)